salt/config/__init__.py

```diff
@@ -840,6 +840,13 @@ def _gather_buffer_space():
     # Always generate minion id in lowercase.
     'minion_id_lowercase': bool,
 
+    # The behavior when generating a minion ID. Can specify 'str' or 'func'.
+    # If 'str' is specified (the default), the 'id' option should be set to
+    # the minion ID or left unspecified for default minion ID generation
+    # behavior. If 'func' is specified, the 'id' option should be set to an
+    # exec module function to run to determine the minion ID.
+    'minion_id_type': str,
```

Comment:
This is confusing. If that supposed to be a function from a module, then I'd call it as `minion_id_function` or something. And the interface to the user would be something like:
```yaml
minion_id_function: salt.modules.mymodule.myfunction
```
In the example above you would not need to have two options in config: this one and `id` as an actual function. An absence of this parameter would imply `str` and thus generate an id as usual.